### PR TITLE
Improve minmax_count_projection

### PR DESCRIPTION
**Pull request by amosbird**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve projection analysis to optimize trivial queries such as `count()`.

Now minmax_count_projection should be almost as good as trivial count optimization. And it covers more cases.

This is for https://github.com/ClickHouse/ClickHouse/issues/34779 .

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.h [hunk 1]
```diff
@@ -393,6 +393,7 @@ class MergeTreeData : public IStorage, public WithMutableContext
     Block getMinMaxCountProjectionBlock(
         const StorageMetadataPtr & metadata_snapshot,
         const Names & required_columns,
+        bool has_filter,
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeData.h:396]: It is not clear what this flag is needed for. Could you please update a comment above?

Comment (amosbird) [reply src/Storages/MergeTree/MergeTreeData.h:396]: Done

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 2]
```diff
@@ -4988,22 +5053,28 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
...
                     aggregates.insert(*column);
-                else
+                    continue;
+                }
+
+                if (normalize_count_not_null && dynamic_cast<const AggregateFunctionCount *>(aggregate.function.get()))
                 {
-                    match = false;
-                    break;
+                    const auto * count_column = sample_block.findByName("count()");
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeData.cpp:5068]: I see in the test, there are cases with `count` with arguments, like `count(if(d=4, d, 1))`. But here we try to find a column with name `count()`. So, I don't understand how it works - shouldn't the column name be `count(if(d=4, d, 1))` as well?

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 3]
```diff
@@ -4961,7 +5026,7 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
     };
 
     auto virtual_block = getSampleBlockWithVirtualColumns();
-    auto add_projection_candidate = [&](const ProjectionDescription & projection)
+    auto add_projection_candidate = [&](const ProjectionDescription & projection, bool normalize_count_not_null = false)
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeData.cpp:5029]: This flag name looks mysterious to me. What does it really mean?

**Review by KochetovNicolai — APPROVED** [review]
I cannot say that I do understand every change here, but generally looks ok

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 2 above)
Comment (amosbird) [reply src/Storages/MergeTree/MergeTreeData.cpp:5068]: The `sample_block` is from `minmax_count_projection` which can only have `count()` as the row count column. During projection analysis, all `count(not-null-expression)` will be normalized to `count()` to check if matched.

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 3 above)
Comment (amosbird) [reply src/Storages/MergeTree/MergeTreeData.cpp:5029]: It means not only do we check if `count() == count()`, but also check `count(any not null expression) == count()`

**Pull request merged**